src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -0,0 +1,440 @@
+      BigInteger bigIntegerVal = getDecimalValue(value, type, insertRowsCurrIndex).unscaledValue();
+      stats.addIntValue(bigIntegerVal);
+      bytes = bigIntegerVal.toByteArray();
+      if (bytes.length < length) {
+        byte[] newBytes = new byte[length];
+        Arrays.fill(newBytes, (byte) (bytes[0] < 0 ? -1 : 0));
+        System.arraycopy(bytes, 0, newBytes, length - bytes.length, bytes.length);
+        bytes = newBytes;
+      }
+    }
+    if (bytes != null) {
+      checkFixedLengthByteArray(bytes, length, insertRowsCurrIndex);
+      return bytes;
+    }
+    throw new SFException(
+        ErrorCode.UNKNOWN_DATA_TYPE, logicalTypeAnnotation, type.getPrimitiveTypeName());
+  }
+
+  /**
+   * Parses a group value based on Parquet group logical type.
+   *
+   * @param value value to parse
+   * @param type Parquet column type
+   * @param stats column stats to update
+   * @param defaultTimezone default timezone to use for timestamp parsing
+   * @param insertRowsCurrIndex Used for logging the row of index given in insertRows API
+   * @return list of parsed values
+   */
+  private static ParquetBufferValue getGroupValue(
+      Object value,
+      GroupType type,
+      RowBufferStats stats,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex) {
+    LogicalTypeAnnotation logicalTypeAnnotation = type.getLogicalTypeAnnotation();
+    if (logicalTypeAnnotation == null) {
+      return getStructValue(value, type, stats, defaultTimezone, insertRowsCurrIndex);
+    }
+    if (logicalTypeAnnotation instanceof ListLogicalTypeAnnotation) {
+      return get3LevelListValue(value, type, stats, defaultTimezone, insertRowsCurrIndex);
+    }
+    if (logicalTypeAnnotation instanceof MapLogicalTypeAnnotation) {
+      return get3LevelMapValue(value, type, stats, defaultTimezone, insertRowsCurrIndex);
+    }
+    throw new SFException(
+        ErrorCode.UNKNOWN_DATA_TYPE, logicalTypeAnnotation, type.getClass().getSimpleName());
+  }
+
+  /**
+   * Parses a struct value based on Parquet group logical type. The parsed value is a list of
+   * values, where each element represents a field in the group. For example, an input {@code
+   * {"field1": 1, "field2": 2}} will be parsed as {@code [1, 2]}.
+   */
+  private static ParquetBufferValue getStructValue(
+      Object value,
+      GroupType type,
+      RowBufferStats stats,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex) {
+    Map<String, Object> structVal =
+        DataValidationUtil.validateAndParseIcebergStruct(
+            type.getName(), value, insertRowsCurrIndex);
+    List<Object> listVal = new ArrayList<>(type.getFieldCount());
+    float estimatedParquetSize = ParquetBufferValue.DEFINITION_LEVEL_ENCODING_BYTE_LEN;
+    for (int i = 0; i < type.getFieldCount(); i++) {
+      ParquetBufferValue parsedValue =
+          parseColumnValueToParquet(
+              structVal.getOrDefault(type.getFieldName(i), null),
```

Comment:
Is null a properly handled value for all types of fields..?

Reply:
Afaik, the leaf primitive type will always check if it's nullable. Will add test to verify it.